src/openfermion/utils/_commutator_diagonal_coulomb_operator.py

```diff
@@ -0,0 +1,303 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Faster commutators for two-body operators with diagonal Coulomb terms."""
+
+import warnings
+
+from openfermion import FermionOperator, normal_ordered
+
+
+def commutator_ordered_diagonal_coulomb_with_two_body_operator(
+        operator_a, operator_b, result=None):
+    """Compute the commutator of two-body operators provided that both are
+    normal-ordered and that the first only has diagonal Coulomb interactions.
+
+    Args:
+        operator_a: The first FermionOperator argument of the commutator.
+            All terms must be normal-ordered, and furthermore either hopping
+            operators (i^ j) or diagonal Coulomb operators (i^ i or i^ j^ i j).
+        operator_b: The second FermionOperator argument of the commutator.
+            operator_b can be any arbitrary two-body operator.
+        result (optional): The initial FermionOperator to add to.
+
+    Returns:
+        result: A FermionOperator with the commutator added to it.
+
+    Notes:
+        The function could be readily extended to the case of arbitrary
+        two-body operator_a given that operator_b has the desired form;
+        however, the extra check slows it down without desirable added utility.
+    """
+    if result is None:
+        result = FermionOperator.zero()
+
+    for term_a in operator_a.terms:
+        coeff_a = operator_a.terms[term_a]
+        for term_b in operator_b.terms:
+            coeff_b = operator_b.terms[term_b]
+
+            coefficient = coeff_a * coeff_b
+
+            # If term_a == term_b the terms commute, nothing to add.
+            if term_a == term_b or not term_a or not term_b:
+                continue
+
+            # Case 1: both operators are two-body, operator_a is i^ j^ i j.
+            if (len(term_a) == len(term_b) == 4 and
+                    term_a[0][0] == term_a[2][0] and
+                    term_a[1][0] == term_a[3][0]):
+                _commutator_two_body_diagonal_with_two_body(
+                    term_a, term_b, coefficient, result)
+
+            # Case 2: commutator of a 1-body and a 2-body operator
+            elif (len(term_b) == 4 and len(term_a) == 2) or (
+                    len(term_a) == 4 and len(term_b) == 2):
+                _commutator_one_body_with_two_body(
+                    term_a, term_b, coefficient, result)
+
+            # Case 3: both terms are one-body operators (both length 2)
+            elif len(term_a) == 2 and len(term_b) == 2:
+                _commutator_one_body_with_one_body(
+                    term_a, term_b, coefficient, result)
+
+            # Final case (case 4): violation of the input promise. Still
```

Comment:
I'm a little surprised by the decision to warn the user that their input is out of spec instead of just raise an exception. What's going on with that?

Reply:
My philosophy here is that exceptions are painful and should only be used for fatal issues (something where we really *can't* proceed). Here we can still do the computation, and deliver an alert to the user that things won't be as fast because they broke the input promise.